# Generated by Carl E. McIntosh
# Date: 06 Jan 2015
# This script shows how functional annotation was preformed using Annovar verison
# Annovar version: annovar_2014Nov12
# Annovar Webpage: http://www.openbioinformatics.org/annovar/

##### Directory and Annovar file locations
ANNOVAR_ROOT_DIR=./annovar/annovar_2014Nov12
CONVERT2ANNOVAR=$ANNOVAR_ROOT_DIR/convert2annovar.pl

##### VCF file for the initial 112 Case samples for the KSHV Project.
VCF_FILE=./KSHV.combo.BOTH.recal.vcf

##### Get annotation dbs required for Annovar
$ANNOVAR_ROOT_DIR/annotate_variation.pl -buildver hg19 -downdb snp138 humandb/

##### Convert VCF file to required Annovar input file format
$CONVERT2ANNOVAR -format vcf4 $VCF_FILE --includeinfo -filter pass -allsample -withfreq -comment -dbsnpfile humandb/hg19_snp138.txt -outfile $VCF_FILE.rs.pass.20150106.avinput
# Result
# NOTICE: Finished reading 34385 lines from VCF file
# NOTICE: A total of 26184 locus in VCF file passed QC threshold, representing 24213 SNPs (16634 transitions and 7579 transversions) and 2837 indels/substitutions
# NOTICE: Finished writing allele frequencies based on 2711856 SNP genotypes (1863008 transitions and 848848 transversions) and 317744 indels/substitutions for 112 samples

##### Run Annovar in Gene-based Annotation 
$ANNOVAR_ROOT_DIR/annotate_variation.pl -geneanno -buildver hg19 $VCF_FILE.rs.pass.20150106.avinput \
./static/annovar/annovar_2014Nov12/humandb
# Result
# NOTICE: Reading gene annotation from ./static/annovar/annovar_2014Nov12/humandb/hg19_refGene.txt ... Done with 49665 transcripts (including 10886 without coding sequence annotation) for 25936 unique genes
# NOTICE: Reading FASTA sequences from ./static/annovar/annovar_2014Nov12/humandb/hg19_refGeneMrna.fa ... Done with 235 sequences
# WARNING: A total of 342 sequences will be ignored due to lack of correct ORF annotation
# NOTICE: Finished gene-based annotation on 27185 genetic variants in ./static/KSHV_2015_OCT_ANALYSIS/ANALYSIS_DIR/FINAL_VCFS/KSHV.combo.BOTH.recal.vcf.rs.pass.20150106.avinput (including 135 with invalid format written to ./static/KSHV_2015_OCT_ANALYSIS/ANALYSIS_DIR/FINAL_VCFS/KSHV.combo.BOTH.recal.vcf.rs.pass.20150106.avinput.invalid_input)
# NOTICE: Output files were written to ./static/KSHV_2015_OCT_ANALYSIS/ANALYSIS_DIR/FINAL_VCFS/KSHV.combo.BOTH.recal.vcf.rs.pass.20150106.avinput.variant_function, ./static/KSHV_2015_OCT_ANALYSIS/ANALYSIS_DIR/FINAL_VCFS/KSHV.combo.BOTH.recal.vcf.rs.pass.20150106.avinput.exonic_variant_function

##### Determine the number of nonsynonymous changes
grep nonsynonymous ./static/KSHV_2015_OCT_ANALYSIS/ANALYSIS_DIR/FINAL_VCFS/KSHV.combo.BOTH.recal.vcf.rs.pass.20150106.avinput.exonic_variant_function | wc -l 
# Result
# 325

##### Determine the number of synonymous changes
grep -w synonymous ./static/KSHV_2015_OCT_ANALYSIS/ANALYSIS_DIR/FINAL_VCFS/KSHV.combo.BOTH.recal.vcf.rs.pass.20150106.avinput.exonic_variant_function | wc -l 
# Result
# 336

##### Determine Class of variants and count
sort \
./static/KSHV_2015_OCT_ANALYSIS/ANALYSIS_DIR/FINAL_VCFS/KSHV.combo.BOTH.recal.vcf.rs.pass.20150106.avinput.variant_function \
| awk '{print $1}' | uniq -c
# Result
#     689 downstream
#     681 exonic
#    6196 intergenic
#   16025 intronic
#     142 ncRNA_exonic
#    1536 ncRNA_intronic
#       1 ncRNA_splicing
#       3 splicing
#     194 upstream
#      15 upstream;downstream
#     461 upstream
#     902 UTR3
#     205 UTR5

